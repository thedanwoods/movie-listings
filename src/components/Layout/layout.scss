.layout {
  background-color: #111;
  color: #bbb;
  min-height: 100vh;
  padding: 16px 0;
}

.layout--loading {
  margin: auto;
  grid-column: span 99;
}

@media screen and (min-width: 450px) {
  .layout {
    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-column-gap: 8px;
    padding: 16px;
  }
}

@media screen and (min-width: 600px) {
  .layout {
    grid-template-columns: 1fr 1fr 1fr;
  }
}

@media screen and (min-width: 800px) {
  .layout {
    grid-template-columns: 1fr 1fr 1fr 1fr;
  }
}

@media screen and (min-width: 450px) {
  .layout__sidebar {
    grid-column: 1;
    grid-row: span 99;
  }
  
  .layout__sidebar-inner {
    position: sticky;
    top: 0;
  }
}
