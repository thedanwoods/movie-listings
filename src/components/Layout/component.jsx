import React, { useEffect } from 'react';
import PropTypes from 'prop-types';
import Spinner from 'react-spinkit';
import Movies from '../Movies';
import GenreFilter from '../GenreFilter';
import RatingFilter from '../RatingFilter';

import './layout.scss';

const Layout = ({ loadStatus, getMovies }) => {
  useEffect(() => {
    // Fetch movies on component mount if never loaded
    if (!loadStatus) getMovies();
  }, []);

  return (
    <div className="layout">
      {loadStatus === 'SUCCESS' && (
        <>
          <div className="layout__sidebar">
            <div className="layout__sidebar-inner">
              <GenreFilter />
              <RatingFilter />
            </div>
          </div>
          <Movies />
        </>
      )}
      {loadStatus === 'ERROR' && (
        <div className="layout__error">
          Sorry, we couldn&apos;t retrieve the movie data. Please reload the
          page or check your connection.
        </div>
      )}
      {loadStatus === 'LOADING' && (
        <div className="layout--loading">
          <Spinner name="ball-beat" color="#bbb" />
        </div>
      )}
    </div>
  );
};

Layout.propTypes = {
  loadStatus: PropTypes.string.isRequired,
  getMovies: PropTypes.func.isRequired,
};

export default Layout;
