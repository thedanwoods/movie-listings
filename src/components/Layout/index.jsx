import { connect } from 'react-redux';
import Layout from './component';
import getMovies from '../../actions/getMovies';

export const mapStateToProps = state => ({
  loadStatus: state.fetchStatus.fetchMovies,
});

export const mapDispatchToProps = dispatch => ({
  getMovies: () => {
    dispatch(getMovies());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(Layout);
