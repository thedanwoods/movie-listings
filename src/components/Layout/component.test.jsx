/*
 Using mount rather than shallow as it is not (yet)
 possible to shallow render components with React Hooks
 https://github.com/facebook/react/issues/14091
*/

import React from 'react';
import { mount } from 'enzyme';
import Layout from './component';

jest.mock('react-spinkit');
jest.mock('../Movies');
jest.mock('../GenreFilter');
jest.mock('../RatingFilter');

const getMoviesMock = jest.fn();

const defaultProps = {
  loadStatus: 'SUCCESS',
  getMovies: getMoviesMock,
};

describe('<Layout /> component', () => {
  it('renders without errors', () => {
    const layout = mount(<Layout {...defaultProps} />);

    expect(() => layout).not.toThrow();
    expect(layout.debug()).toMatchSnapshot();
  });

  it('displays an error if loadStatus is "ERROR"', () => {
    const props = {
      ...defaultProps,
      loadStatus: 'ERROR',
    };
    const layout = mount(<Layout {...props} />);
    expect(layout.find('.layout__error').exists()).toBe(true);
    expect(layout.debug()).toMatchSnapshot();
  });

  it('displays a spinner if loadStatus is "LOADING"', () => {
    const props = {
      ...defaultProps,
      loadStatus: 'LOADING',
    };
    const layout = mount(<Layout {...props} />);
    expect(layout.find('Spinner').exists()).toBe(true);
    expect(layout.debug()).toMatchSnapshot();
  });
});
