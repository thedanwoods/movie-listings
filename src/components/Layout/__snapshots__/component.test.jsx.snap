// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Layout /> component displays a spinner if loadStatus is "LOADING" 1`] = `
"<Layout loadStatus=\\"LOADING\\" getMovies={[Function: mockConstructor]}>
  <div className=\\"layout\\">
    <div className=\\"layout--loading\\">
      <Spinner name=\\"ball-beat\\" color=\\"#bbb\\" noFadeIn={false} fadeIn=\\"full\\" overrideSpinnerClassName=\\"\\" />
    </div>
  </div>
</Layout>"
`;

exports[`<Layout /> component displays an error if loadStatus is "ERROR" 1`] = `
"<Layout loadStatus=\\"ERROR\\" getMovies={[Function: mockConstructor]}>
  <div className=\\"layout\\">
    <div className=\\"layout__error\\">
      Sorry, we couldn&#39;t retrieve the movie data. Please reload the page or check your connection.
    </div>
  </div>
</Layout>"
`;

exports[`<Layout /> component renders without errors 1`] = `
"<Layout loadStatus=\\"SUCCESS\\" getMovies={[Function: mockConstructor]}>
  <div className=\\"layout\\">
    <div className=\\"layout__sidebar\\">
      <div className=\\"layout__sidebar-inner\\">
        <Connect(GenreFilter) />
        <Connect(RatingFilter) />
      </div>
    </div>
    <Connect(Movies) />
  </div>
</Layout>"
`;
